src/charm.py

```diff
@@ -589,21 +593,36 @@ def _update_cert(self):
                 self.server_cert.ca_cert,  # pyright: ignore
                 make_dirs=True,
             )
+        except ConnectionError:
+            logger.error("Could not push cert files. Pebble refused connection. Shutting down?")
+            return
 
-            # Repeat for the charm container. We need it there for loki client requests.
-            # (and charm tracing and logging TLS)
-            ca_cert_path.parent.mkdir(exist_ok=True, parents=True)
-            ca_cert_path.write_text(self.server_cert.ca_cert)  # pyright: ignore
-        else:
+        # Repeat for the charm container. We need it there for loki client requests.
+        # (and charm tracing and logging TLS)
+        ca_cert_path.parent.mkdir(exist_ok=True, parents=True)
+        ca_cert_path.write_text(self.server_cert.ca_cert)  # pyright: ignore
+
+    def _remove_cert_files(self):
+        ca_cert_path = Path(self._ca_cert_path)
+        try:
             self._loki_container.remove_path(CERT_FILE, recursive=True)
             self._loki_container.remove_path(KEY_FILE, recursive=True)
             self._loki_container.remove_path(ca_cert_path, recursive=True)
+        except ConnectionError:
+            logger.error("Could not remove cert files. Pebble refused connection. Shutting down?")
+            return
```

Comment:
I think we should not return here but surface the error. If we are shutting down, then it doesn't really matter because the charm will go away soon. If we are not shutting down though, these files will stay on disk forever.

Reply:
We can avoid returning here and let the workflow continue.
This exception is raised if pebble is not ready. 

We execute `_update_cert()` which conditionally executes `_remove_cert_files()` on:

- `__init__` method
- `server_cert.on.cert_changed` event handler
- `_configure` common exit hook method

In order to avoid this files stay on disk forever, we can execute this method before anything else:

```diff
     def _update_cert(self):
+        self._remove_cert_files()
+
         if self.server_cert.available:
             self._push_cert_files()
-        else:
-            self._remove_cert_files()
 
         self._update_ca_certificates()

```

WDYT?

Reply:
I am not sure. I think I would still prefer to surface the exception because another hook might not fire. I don't feel strongly on this one though so it is in no way blocking my approval.